closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -1661,30 +1763,55 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_RSADP
     msg_len = (*OSSL_RSA_private_encrypt)(kLen, kNative, mNative, rsaKey, RSA_NO_PADDING);
 
     if ((-1 != verify) && (-1 != msg_len)) {
-        if (verify == kLen) {
-            k2 = malloc(kLen * (sizeof(unsigned char)));
-            if (NULL != k2) {
-
-                /* mNative is size 'verify' */
-                msg_len2 = (*OSSL_RSA_public_decrypt)(verify, mNative, k2, rsaKey, RSA_NO_PADDING);
-                if (-1 != msg_len2) {
-
-                    int i;
-                    for (i = 0; i < verify; i++) {
-                        if (kNative[i] != k2[i]) {
-                            msg_len = -2;
-                            break;
+        if (verify == msg_len) {
+            if(-1 != rsadpBNptr) {
+                rsadp_bn_ptr = (RSADP_BN *)rsadpBNptr;
+                /* We are using OpenSSL 1.1.*
+                 * Verify by comparing BigNum of mNative and m_bigNum ^ e % n */
+                const BIGNUM *n = NULL;
+                const BIGNUM *e = NULL;
+                (*OSSL_RSA_get0_key)(rsaKey, &n, &e, NULL);
+                (*OSSL_BN_bin2bn)(kNative, kLen, rsadp_bn_ptr->k_bigNum);
+                (*OSSL_BN_bin2bn)(mNative, msg_len, rsadp_bn_ptr->m_bigNum);
+                if (NULL != rsadp_bn_ptr->k_bigNum && NULL != rsadp_bn_ptr->m_bigNum && NULL != e && NULL != n) {
+                    if (!(*OSSL_BN_mod_exp)(rsadp_bn_ptr->rem, rsadp_bn_ptr->m_bigNum, e, n, rsadp_bn_ptr->ctx)
+                        || (0 != (*OSSL_BN_ucmp)(rsadp_bn_ptr->k_bigNum, rsadp_bn_ptr->rem))
+                    ) {
+                        msg_len = -1;
+                    }
+                } else {
+                    msg_len = -1;
+                }
+                (*OSSL_BN_clear)(rsadp_bn_ptr->k_bigNum);
+                (*OSSL_BN_clear)(rsadp_bn_ptr->m_bigNum);
+                (*OSSL_BN_clear)(rsadp_bn_ptr->rem);
+            }
+            else {
+                /* We are using OpenSSL 1.0.*
+                 * Verify by comparing RSA_public_decrypt with original msg
+                */
+                k2 = malloc(kLen * (sizeof(unsigned char)));
+                if (NULL != k2) {
+                    /* mNative is size 'verify' */
+                    msg_len2 = (*OSSL_RSA_public_decrypt)(verify, mNative, k2, rsaKey, RSA_NO_PADDING);
```

Comment:
Why doesn't this approach also work for OpenSSL 1.1.*?

Reply:
I still don't understand why 1025 is special. Please explain, and make sure the answer is in the code somewhere for easy future reference.

Reply:
The test is only failing when the key size is `8n + 1` (where n is a whole number). I tried running the test (when triggering OpenSSL's `RSA_public_decrypt`) on key sizes: 1023, 1024, 1025, and 1026; the test is only failing on 1025.

Hence, I updated the code to check for that pattern (`keySize % 8 == 1`).